src/lib/user-performance-metrics.ts

```diff
@@ -0,0 +1,138 @@
+// Forked and modified from UXM: https://github.com/treosh/uxm
+
+declare let PerformancePaintTiming: any
+
+const perf = typeof window !== "undefined" ? window.performance : null
+
+const sanitizedMetrics = (start, end) =>
+  start > 0 && end > 0 && end - start > 0 ? end - start : null
+
+// user timing helpers
+
+export function mark(markName) {
+  if (perf && perf.mark) {
+    perf.mark(markName)
+  }
+}
+
+export function measure(measureName, startMarkName) {
+  if (perf && perf.measure) {
+    try {
+      perf.measure(measureName, startMarkName)
+    } catch (err) {
+      console.error(err)
+    }
+  }
+}
+
+// default metrics
+
+// export function getEffectiveConnectionType() {
+//   const conn =
+//     typeof navigator !== "undefined"
+//       ? navigator.connection ||
+//         navigator.mozConnection ||
+//         navigator.webkitConnection
+//       : null
+//   return conn ? conn.effectiveType : null
+// }
+
+export function getFirstPaint() {
+  if (typeof PerformancePaintTiming === "undefined") return null
```

Comment:
Do we really need to test `PerformancePaintTiming` instead of `perf !== null`?

Reply:
Yeah. That's a feature flag for that specific measurement. If it isn't present then there will be no paint events captured and if it is we know the performance interface exists because it's a subset of that system.

Reply:
Ah gotcha 👍